#### Parte 1: Carga y Exploración de Datos

**Carga del Conjunto de Datos**

Instalar dependencias

In [18]:
%pip install pandas
%pip install db-dtypes


Note: you may need to restart the kernel to use updated packages.


Importar pandas e importar el dataset

In [19]:
import pandas as pd

data_path="../data/datos.csv"

df =pd.read_csv(data_path)
df

productAsin country        date  isVerified  ratingScore  \
0     B09G9BL5CP   India  11-08-2024        True            4   
1     B09G9BL5CP   India  16-08-2024        True            5   
2     B09G9BL5CP   India  14-05-2024        True            4   
3     B09G9BL5CP   India  24-06-2024        True            5   
4     B09G9BL5CP   India  18-05-2024        True            5   
...          ...     ...         ...         ...          ...   
3057  B09G9D8KRQ   India  11-10-2023        True            1   
3058  B09G9D8KRQ   India  14-10-2022        True            2   
3059  B09G9D8KRQ   India  24-02-2022        True            3   
3060  B09G9D8KRQ   India  16-10-2023        True            1   
3061  B09G9D8KRQ   India  11-11-2023        True            1   

                              reviewTitle  \
0                              No charger   
1                         iPhone 13 256GB   
2                 Flip camera option nill   
3                                 Product   
4                            Good product   
...                                   ...   
3057  Very bad experience with i phone 13   
3058    not happy with this apple product   
3059                           Good phone   
3060                    Battery discharge   
3061    Batter power needs to be improved   

                                      reviewDescription  \
0     Every thing is good about iPhones, there's not...   
1     It look so fabulous, I am android user switche...   
2     I tried to flip camera while recording but no ...   
3                                          100% genuine   
4            Happy to get the iPhone 13 in Amazon offer   
...                                                 ...   
3057  Useless phon never buy this heat n useless cam...   
3058  iam not happy with this product why because ch...   
3059                                         Good phone   
3060  While charging mobile it's getting so hot even...   
3061  Battery power is be very bad need to chat on d...   

                                              reviewUrl  \
0     https://www.amazon.in/gp/customer-reviews/R345...   
1     https://www.amazon.in/gp/customer-reviews/R2HJ...   
2     https://www.amazon.in/gp/customer-reviews/R3Y7...   
3     https://www.amazon.in/gp/customer-reviews/R1P9...   
4     https://www.amazon.in/gp/customer-reviews/R1XI...   
...                                                 ...   
3057  https://www.amazon.in/gp/customer-reviews/R10O...   
3058  https://www.amazon.in/gp/customer-reviews/R2FW...   
3059  https://www.amazon.in/gp/customer-reviews/R2C7...   
3060  https://www.amazon.in/gp/customer-reviews/R3K0...   
3061  https://www.amazon.in/gp/customer-reviews/R2QO...   

                                 reviewedIn  \
0       Reviewed in India on 11 August 2024   
1       Reviewed in India on 16 August 2024   
2          Reviewed in India on 14 May 2024   
3         Reviewed in India on 24 June 2024   
4          Reviewed in India on 18 May 2024   
...                                     ...   
3057   Reviewed in India on 11 October 2023   
3058   Reviewed in India on 14 October 2022   
3059  Reviewed in India on 24 February 2022   
3060   Reviewed in India on 16 October 2023   
3061  Reviewed in India on 11 November 2023   

                                variant variantAsin  
0          Colour: MidnightSize: 256 GB  B09G9BQS98  
1          Colour: MidnightSize: 256 GB  B09G9BQS98  
2          Colour: MidnightSize: 256 GB  B09G9BQS98  
3          Colour: MidnightSize: 256 GB  B09G9BQS98  
4          Colour: MidnightSize: 256 GB  B09G9BQS98  
...                                 ...         ...  
3057  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  
3058  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  
3059  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  
3060  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  
3061  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  

[3062 rows x 11 columns]

Se identifican las columnas reviewDescription que contiene el texto de la reseña y ratingScore que indica la valoracion, segun la valoracion podemos identificar si la reseña es positiva o negativa.

**Descripción Inicial de los Datos**

Se cuenta la cantidad de reseñas presentes

In [20]:
num_filas = df.shape[0]
print(f"El DataFrame tiene {num_filas} filas.")

El DataFrame tiene 3062 filas.


El dataframe tiene  3062 reseñas

Ahora veremos si hay mas reseñas positivas o negativas. En nuestro caso, basado en la columna **ratingScore**, positivo sera mayor que 3 y menor o igual a 3 sera negativo. Para diferenciarlo mejor, añadiremos una columna booleana que contendra un *Positive* o *Negative*.

In [21]:
df['ratingPositive'] = df['ratingScore'].apply(lambda x: 'Positive' if x > 3 else 'Negative')

df.to_csv("../data/datosConColumnaPositivos.csv",index=False)
df

productAsin country        date  isVerified  ratingScore  \
0     B09G9BL5CP   India  11-08-2024        True            4   
1     B09G9BL5CP   India  16-08-2024        True            5   
2     B09G9BL5CP   India  14-05-2024        True            4   
3     B09G9BL5CP   India  24-06-2024        True            5   
4     B09G9BL5CP   India  18-05-2024        True            5   
...          ...     ...         ...         ...          ...   
3057  B09G9D8KRQ   India  11-10-2023        True            1   
3058  B09G9D8KRQ   India  14-10-2022        True            2   
3059  B09G9D8KRQ   India  24-02-2022        True            3   
3060  B09G9D8KRQ   India  16-10-2023        True            1   
3061  B09G9D8KRQ   India  11-11-2023        True            1   

                              reviewTitle  \
0                              No charger   
1                         iPhone 13 256GB   
2                 Flip camera option nill   
3                                 Product   
4                            Good product   
...                                   ...   
3057  Very bad experience with i phone 13   
3058    not happy with this apple product   
3059                           Good phone   
3060                    Battery discharge   
3061    Batter power needs to be improved   

                                      reviewDescription  \
0     Every thing is good about iPhones, there's not...   
1     It look so fabulous, I am android user switche...   
2     I tried to flip camera while recording but no ...   
3                                          100% genuine   
4            Happy to get the iPhone 13 in Amazon offer   
...                                                 ...   
3057  Useless phon never buy this heat n useless cam...   
3058  iam not happy with this product why because ch...   
3059                                         Good phone   
3060  While charging mobile it's getting so hot even...   
3061  Battery power is be very bad need to chat on d...   

                                              reviewUrl  \
0     https://www.amazon.in/gp/customer-reviews/R345...   
1     https://www.amazon.in/gp/customer-reviews/R2HJ...   
2     https://www.amazon.in/gp/customer-reviews/R3Y7...   
3     https://www.amazon.in/gp/customer-reviews/R1P9...   
4     https://www.amazon.in/gp/customer-reviews/R1XI...   
...                                                 ...   
3057  https://www.amazon.in/gp/customer-reviews/R10O...   
3058  https://www.amazon.in/gp/customer-reviews/R2FW...   
3059  https://www.amazon.in/gp/customer-reviews/R2C7...   
3060  https://www.amazon.in/gp/customer-reviews/R3K0...   
3061  https://www.amazon.in/gp/customer-reviews/R2QO...   

                                 reviewedIn  \
0       Reviewed in India on 11 August 2024   
1       Reviewed in India on 16 August 2024   
2          Reviewed in India on 14 May 2024   
3         Reviewed in India on 24 June 2024   
4          Reviewed in India on 18 May 2024   
...                                     ...   
3057   Reviewed in India on 11 October 2023   
3058   Reviewed in India on 14 October 2022   
3059  Reviewed in India on 24 February 2022   
3060   Reviewed in India on 16 October 2023   
3061  Reviewed in India on 11 November 2023   

                                variant variantAsin ratingPositive  
0          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
1          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
2          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
3          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
4          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
...                                 ...         ...            ...  
3057  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative  
3058  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative  
3059  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative  
3060  Colo

Contamos la cantidad de *Positive* y *Negative*

In [22]:
conteo_positivos_negativos = df['ratingPositive'].value_counts()

print(conteo_positivos_negativos)

ratingPositive
Positive    2065
Negative     997
Name: count, dtype: int64


Hay 2065 reseñas positivas y 997 negativas.

#### Parte 2: Preprocesamiento de Texto

**Preprocesamiento del Texto**

Instalar spacy

In [23]:
%pip install spacy
!python -m spacy download en_core_web_sm

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     ---------------------------- ----------- 9.2/12.8 MB 57.1 MB/s eta 0:00:01
     --------------------------------------- 12.8/12.8 MB 53.6 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


Tokenización  
Ademas se quitan espacios vacios y caracteres molestos como "," y "."

In [24]:
import spacy

nlp = spacy.load("en_core_web_sm")

df['reviewDescription'] = df['reviewDescription'].fillna('')

df['preprocessedReview'] = ''

df['preprocessedReview'] = df['reviewDescription'].apply(lambda doc: [token.text for token in nlp(doc) if not token.is_punct and not token.is_space and token.text not in ['"', "'"]])

Quitar Stop Words

In [25]:
df['preprocessedReview'] = df['preprocessedReview'].apply(lambda tokens: [word for word in tokens if not nlp.vocab[word].is_stop])

Lematización

In [31]:
df['preprocessedReview'] = df['preprocessedReview'].apply(lambda tokens: [token.lemma_ for token in nlp(" ".join(tokens))])

In [32]:
df.to_csv("../data/a.csv",index=False)

Las técnicas aplicadas ayudan a limpiar y preparar el texto de la siguiente manera:  
**Tokenización**: Divide el texto en palabras individuales (tokens), facilitando el análisis palabra por palabra. Esto permite procesar el texto de manera más estructurada.  
**Eliminación de Stop Words**: Remueve palabras comunes que no aportan mucho significado (como "the", "and", etc.), permitiendo que el análisis se concentre en palabras más relevantes.  
**Lematización**: Reduce las palabras a su forma base o raíz, eliminando variaciones gramaticales (por ejemplo, "running" se convierte en "run"). Esto ayuda a unificar conceptos y mejora la precisión en el análisis.  
Estas técnicas son esenciales para eliminar ruido en el texto y convertirlo en una forma más adecuada para el análisis posterior, ademas de reducir el tiempo de lectura y ejecucion del codigo.


#### Parte 3: Representación del Texto con Bag of Words

**Bag of Words (BoW)**

Instalar scikit-learn

In [34]:
%pip install scikit-learn


   ---------------------------------------- 0.0/11.0 MB ? eta -:--:--
   --------------------------- ------------ 7.6/11.0 MB 42.9 MB/s eta 0:00:01
   ---------------------------------------- 11.0/11.0 MB 42.9 MB/s eta 0:00:00
   ---------------------------------------- 0.0/44.5 MB ? eta -:--:--
   ------ --------------------------------- 6.8/44.5 MB 38.1 MB/s eta 0:00:01
   ------------- -------------------------- 14.7/44.5 MB 35.5 MB/s eta 0:00:01
   -------------------- ------------------- 23.3/44.5 MB 38.8 MB/s eta 0:00:01
   ----------------------------- ---------- 32.5/44.5 MB 40.5 MB/s eta 0:00:01
   ------------------------------------- -- 41.9/44.5 MB 41.1 MB/s eta 0:00:01
   ---------------------------------------- 44.5/44.5 MB 39.3 MB/s eta 0:00:00


In [47]:
from sklearn.feature_extraction.text import CountVectorizer

df['preprocessedReviewJoined'] = df['preprocessedReview'].apply(lambda tokens: ' '.join(tokens))

vectorizer = CountVectorizer()

X_bow = vectorizer.fit_transform(df['preprocessedReviewJoined'])

bow_df = pd.DataFrame(X_bow.toarray(), columns=vectorizer.get_feature_names_out())


In [48]:
bow_df.to_csv("../data/bow.csv",index=False)

Bag of Words (BoW) basicamente convirtio las reseñas preprocesadas (columna preprocessedReview) en vectores numéricos. Cada reseña es transformada en una fila de la matriz, donde cada columna representa una palabra del vocabulario total de las reseñas. El valor en cada celda indica cuántas veces aparece esa palabra en la reseña específica.
BoW es importante porque convierte texto en una representación numérica, es simple porque solo cuenta la frecuencia de palabras sin considerar el orden, y es eficaz al crear una matriz numérica fácil de interpretar y procesar por algoritmos de machine learning.

#### Parte 4: Implementación de Clasificadores

**Clasificador Naive Bayes**

In [49]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

X_train, X_test, y_train, y_test = train_test_split(X_bow, df['ratingPositive'], test_size=0.2, random_state=42)

nb_classifier = MultinomialNB()

nb_classifier.fit(X_train, y_train)

y_pred = nb_classifier.predict(X_test)


Precisión del modelo Naive Bayes: 0.86


Evaluar rendimiento

In [50]:
from sklearn.metrics import classification_report

report = classification_report(y_test, y_pred, target_names=['Negative', 'Positive'])

print(report)


              precision    recall  f1-score   support

    Negative       0.81      0.70      0.75       183
    Positive       0.88      0.93      0.90       430

    accuracy                           0.86       613
   macro avg       0.84      0.81      0.83       613
weighted avg       0.86      0.86      0.86       613



Al parecer el modelo es más preciso al predecir reseñas positivas (88%) que negativas (81%). por otra paerte tiene un buen recall de 93% para las reseñas positivas, a diferencia de un recall menor para las reseñas negativas (70%).
El modelo tiene una precisión general del 86% seun el accuracy.En general, el modelo tiene un mejor desempeño en las reseñas positivas que en las negativas. Esto puede deberse a un mayor número de reseñas positivas en el conjunto de datos (430 frente a 183).

**Regresión Logística**

In [51]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# Usar la misma matriz de características (BoW) y las etiquetas
X = X_bow  # Matriz BoW
y = df['ratingPositive']  # Etiquetas de la columna 'ratingPositive'

# Dividir los datos en conjunto de entrenamiento y prueba (70% entrenamiento, 30% prueba)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Crear un modelo de Regresión Logística
logistic_model = LogisticRegression(max_iter=1000)

# Entrenar el modelo con los datos de entrenamiento
logistic_model.fit(X_train, y_train)

# Hacer predicciones en el conjunto de prueba
y_pred = logistic_model.predict(X_test)

# Evaluar el rendimiento del modelo utilizando el informe de clasificación
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

    Negative       0.86      0.66      0.75       281
    Positive       0.86      0.95      0.91       638

    accuracy                           0.86       919
   macro avg       0.86      0.81      0.83       919
weighted avg       0.86      0.86      0.86       919



El modelo de Regresión Logística muestra un rendimiento similar al de Naive Bayes, con una precisión del 86% para ambas clases, positivas y negativas. Sin embargo, tiene un mejor recall para las reseñas positivas (95%) en comparación con las negativas (66%), lo que sugiere que identifica mejor las reseñas positivas. El F1-score también refleja un mejor balance para las reseñas positivas (91%) frente al 75% para las negativas. En general, el modelo tiene una precisión global del 86% y, al igual que Naive Bayes, se desempeña mejor en la clasificación de reseñas positivas, lo que probablemente se deba a la mayor cantidad de ejemplos positivos (638 frente a 281 negativos) en el conjunto de datos.

#### Parte 5: Análisis y Conclusión

**Análisis de Resultados**

Comparando los resultados de Naive Bayes y Regresión Logística:  

**Precisión**:  
Ambos modelos tienen precisión similar. En el caso de los valores negativos, el modelo de regresión logística tiene una precisión ligeramente mejor (0.86 frente a 0.81).  

Para los valores positivos, ambos modelos tienen la misma precisión (0.86-0.88).  

**Recall**:  
En la categoría negativa, Naive Bayes tiene mejor recall (0.70 frente a 0.66).  

Para los valores positivos, la regresión logística tiene mejor recall (0.95 frente a 0.93), lo que significa que identifica más correctamente las instancias positivas.  

**F1-Score**:  
El F1-Score, que combina la precisión y el recall, es ligeramente mejor para la regresión logística en los valores positivos (0.91 frente a 0.90). Por otra parte, en los valores negativos, tienen igual desempeño (0.75 frente a 0.75).  

**Conclusión**:  
Naive Bayes es más equilibrado en identificar tanto negativos como positivos, aunque tiene un menor recall para positivos.
Regresión Logística sobresale en identificar correctamente los positivos con un recall más alto, pero pierde algo de precisión en la clase negativa.  

En general, Regresión Logística parece ser mejor para maximizar la detección de positivos, mientras que Naive Bayes es más equilibrado en ambas clases.

**Métodos de mejora sugeridos**

Mejor Preprocesamiento  

1. Normalización o Escalado: Los modelos como la regresión logística suelen beneficiarse de la normalización de los datos. Escalar las características podría mejorar su desempeño.  
2. Textos Limpios y Consistentes: Considerar el uso de técnicas más avanzadas para limpiar los textos, como eliminar palabras muy comunes y no informativas.  
3. N-gramas: Usar n-gramas (combinaciones de palabras) en lugar de palabras individuales podría capturar mejor el contexto en las reseñas, lo que puede mejorar el modelo.  

Mejoras en el Modelo:

1. Hiperparámetros: Ajustar hiperparámetros, como la regularización en la regresión logística o suavizado en Naive Bayes, podría mejorar los resultados.
2. Ensamble de Modelos: Se podría explorar un ensamble de ambos modelos (Naive Bayes + Regresión Logística) para aprovechar las fortalezas de cada uno.
3. Modelos más avanzados: Explorar otros algoritmos de clasificación, como máquinas de soporte vectorial (SVM) o redes neuronales, podría mejorar el desempeño.
